sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala

```diff
@@ -0,0 +1,304 @@
+
+    assertOnSparkDriver()
+    cache.get(key) match {
+      case Some(metadata) =>
+        metadata.lastRetrieved.set(Instant.now.toEpochMilli)
+        SMono.just(metadata)
+      case None =>
+        val metadataObservable = readPartitionMetadata(
+          cosmosClientConfiguration: CosmosClientConfiguration,
+          cosmosClientStateHandle: Option[Broadcast[CosmosClientMetadataCachesSnapshot]],
+          cosmosContainerConfig: CosmosContainerConfig,
+          feedRange: String
+        )
+
+        metadataObservable
+          .map(metadata => {
+            cache.putIfAbsent(key, metadata) match {
+              case None =>
+                metadata
+              case Some(metadataAddedConcurrently) =>
+                metadataAddedConcurrently.lastRetrieved.set(Instant.now.toEpochMilli)
+
+                metadataAddedConcurrently
+            }
+          })
+          .subscribeOn(Schedulers.boundedElastic())
+    }
+  }
+
+  private def readPartitionMetadata
+  (
+    cosmosClientConfiguration: CosmosClientConfiguration,
+    cosmosClientStateHandle: Option[Broadcast[CosmosClientMetadataCachesSnapshot]],
+    cosmosContainerConfig: CosmosContainerConfig,
+    feedRange: String
+  ): SMono[PartitionMetadata] = {
+    val client = CosmosClientCache.apply(cosmosClientConfiguration, cosmosClientStateHandle)
+    val container = client
+      .getDatabase(cosmosContainerConfig.database)
+      .getContainer(cosmosContainerConfig.container)
+
+    val options = CosmosChangeFeedRequestOptions.createForProcessingFromNow(FeedRange.fromString(feedRange))
+    options.setMaxItemCount(1)
+    options.setMaxPrefetchPageCount(1)
+    options.setQuotaInfoEnabled(true)
+
+    val lastDocumentCount = new AtomicLong()
+    val lastTotalDocumentSize = new AtomicLong()
+    val lastContinuationToken = new AtomicReference[String]()
+
+    container
+      .queryChangeFeed(options, classOf[ObjectNode])
+      .handle(r => {
+        lastDocumentCount.set(r.getDocumentCountUsage)
+        lastTotalDocumentSize.set(r.getDocumentUsage)
+        val continuation = r.getContinuationToken
+        if (!Strings.isNullOrWhiteSpace(continuation)) {
+          lastContinuationToken.set(continuation)
+        }
+      })
+      .collectList()
+      .asScala
+      .map(_ => {
+        PartitionMetadata(
+          cosmosClientConfiguration,
+          cosmosClientStateHandle,
+          cosmosContainerConfig,
+          feedRange,
+          assertNotNull(lastDocumentCount.get, "lastDocumentCount"),
+          assertNotNull(lastTotalDocumentSize.get, "lastTotalDocumentSize"),
+          assertNotNullOrEmpty(lastContinuationToken.get, "continuationToken")
+        )
+      })
+  }
+
+  private def startRefreshTimer() : Unit = {
+    logInfo(s"$timerName: scheduling timer - delay: $refreshIntervalInMs ms, period: $refreshIntervalInMs ms")
+    testTimerOverride.getOrElse(timer).schedule(
+      new TimerTask { def run(): Unit = onRunRefreshTimer() },
```

Comment:
ditto
TimerTask is a very old concurrency api in java 1.3.
any reason not to use `ScheduledThreadExecutor` or `ExecutorService`

using `ScheduledThreadPoolExecutor` is usually preferred as you would have better control over the threading, etc.
https://stackoverflow.com/questions/409932/java-timer-vs-executorservice

Reply:
See above